util/argo/argo.go

```diff
@@ -508,6 +508,17 @@ func ValidateDestination(ctx context.Context, dest *argoappv1.ApplicationDestina
 		} else if !dest.IsServerInferred() {
 			return fmt.Errorf("application destination can't have both name and server defined: %s %s", dest.Name, dest.Server)
 		}
+	} else if dest.Server != "" {
+		if dest.Name == "" {
+			serverName, err := getDestinationServerName(ctx, db, dest.Server)
+			if err != nil {
+				return fmt.Errorf("unable to find destination server: %w", err)
+			}
+			if serverName == "" {
+				return fmt.Errorf("application references destination cluster %s which does not exist", dest.Server)
+			}
+			dest.SetInferredServer(serverName)
```

Comment:
Somewhere you need to also set the dest.Name to the serverName value or else it won't be set.

Reply:
Yup, addressed in via https://github.com/argoproj/argo-cd/pull/21063/commits/8eb2e4ad8a42bd155d9af543a2157bc618c42c56